src/braket/default_simulator/openqasm/program_context.py

```diff
@@ -571,6 +579,44 @@ def add_builtin_gate(
                 ctrl_modifiers += [ctrl_mod_ix] * mod.argument.value
             if mod.modifier == GateModifierName.pow:
                 power *= mod.argument.value
+        self.add_gate_instruction(
+            gate_name, target, params, ctrl_modifiers=ctrl_modifiers, power=power
+        )
+
+    @abstractmethod
+    def add_gate_instruction(
+        self, gate_name: str, target: Tuple[int], params, ctrl_modifiers: List[int], power: int
+    ):
+        """Add gate instruction to the circuit"""
+
+    @abstractmethod
+    def add_custom_unitary(
+        self,
+        unitary: np.ndarray,
+        target: Tuple[int],
+    ) -> None:
+        """Add a custom Unitary instruction to the circuit"""
+
+    @abstractmethod
+    def add_noise_instruction(self, *args, **kwargs):
+        """Add a noise instruction to the circuit"""
+
+
+class ProgramContext(AbstractProgramContext):
+    def __init__(self):
+        super().__init__(Circuit())
```

Comment:
... And set the circuit here
```suggestion
    def __init__(self):
        super().__init__()
        self._circuit = Circuit()
```

Reply:
This behavior is likely duplicated in all subclasses, so to me it would make sense to keep that assignment in the base class

Reply:
Is that necessarily the case though? I can totally conceive of a situation where an implementation looks like
```python
class MyContext(AbstractProgramContext):
    def __init__(self, ...):
        self._a = ...
        self._b = ...

    def exec(self):
        self._a.do_something()
        self._b.do_something()
```

Reply:
The interpreter class does have a `build_circuit` function, which returns `context.circuit`, but other than that there is no direct coupling between the interpreter and circuit classes. If someone wanted to build an interpreter/simulator that didn't have a build_circuit method, then yup I agree

Reply:
Ah, in that case, it does make sense then.

Reply:
Arguably, if we carry on this line of thought, an abstract interpreter class wouldn't have a build_circuit class. For example the mcm interpreter doesn't utilize that method at all, but has a [simulate](https://github.com/aws/amazon-braket-default-simulator-python/blob/mcm-sim/src/braket/default_simulator/openqasm/native_interpreter.py#L39) method instead

Reply:
Looking at `Interpreter`, it seems pretty deeply linked to the default simulator's `Circuit` class; no need to do it in this PR, but it seems to make sense to the visitor functionality from the build/run functionality.